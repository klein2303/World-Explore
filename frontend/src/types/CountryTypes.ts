export type CountryType = {
    name: string;
    continent: string;
    capital: string;
    largest_city: string;
    currency: string;
    language: string;
    population: number;
    land_area: number;
    agriculture_area: string;
    forest_area: string;
    co2_emissions: number;
    image: string;
};
