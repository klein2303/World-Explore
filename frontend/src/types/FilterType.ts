export type FilterType = {
    continent: ContinentType;
    search: string;
    sort: string;
};

type ContinentType = {
    Africa: boolean;
    Asia: boolean;
    Europe: boolean;
    "North America": boolean;
    Oceania: boolean;
    "South America": boolean;
}