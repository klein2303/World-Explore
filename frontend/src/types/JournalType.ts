export type JournalTypeRead = {
    country: string;
    reviews: readReviewType[];
};

export type JournalTypeWrite = {
    country: string;
    reviews: writeReviewType[];
};

export type readReviewType = {
    title: string;
    date: string;
    rating: number;
    text: string;
    public: boolean;
    journal: {
        profile: {
            username: string;
        };
    };
};

export type writeReviewType = {
    title: string;
    date: string;
    rating: number;
    text: string;
    public: boolean;
    journal: {
        profile: {
            email: string;
        };
    };
}