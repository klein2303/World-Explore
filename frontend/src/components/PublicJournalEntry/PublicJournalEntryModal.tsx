import { useEffect } from "react";
import styles from "./PublicJournalEntryModal.module.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faStar } from "@fortawesome/free-solid-svg-icons";
import { CgProfile } from "react-icons/cg";
import { reviewType } from "../../types/JournalType";

type JournalEntryModalProps = {
    review: reviewType;
    onClose: () => void;
};

const PublicJournalEntryModal = ({ review, onClose }: JournalEntryModalProps) => {
    // Close modal on "Escape" key press
    useEffect(() => {
        const handleKeyDown = (event: KeyboardEvent) => {
            if (event.key === "Escape") {
                onClose();
            }
        };
        document.addEventListener("keydown", handleKeyDown);
        return () => document.removeEventListener("keydown", handleKeyDown);
    }, [onClose]);

    return (
        <aside
            className={styles.modalBackdrop}
            onClick={onClose}
            role="dialog"
            aria-modal="true"
            aria-labelledby={`review-title-${review.title}`}
            aria-describedby={`review-text-${review.text}`}>
            <div className={styles.modalContent} onClick={(e) => e.stopPropagation()} tabIndex={-1}>
                {/* Review title, date and close button */}
                <header>
                    <button onClick={onClose} className={styles.closeButton} aria-label="Close modal">
                        &times;
                    </button>
                    <h3 id={`review-title-${review.title}`} className={styles.modalTitle}>
                        {review.title}
                    </h3>
                    <p className={styles.modalDate} aria-label={`Date range: ${review.date}`}>
                        {review.date}
                    </p>
                </header>

                {/* Review and reviewer info */}
                <section className={styles.modalInfo} aria-labelledby="modal-rating modal-reviewer">
                    <p
                        id="modal-rating"
                        className={styles.modalRating}
                        aria-label={`Rating: ${review.rating} out of 5 stars`}>
                        <FontAwesomeIcon icon={faStar} className={styles.starIcon} aria-hidden="true" />
                        {review.rating}/5
                    </p>
                    <p id="modal-reviewer" className={styles.modalReviewer} aria-label="Reviewer: Ola Nordmann">
                        <CgProfile className={styles.profileIcon} aria-hidden="true" />
                        Ola Nordmann
                    </p>
                </section>
                {/* Review text */}
                <article id={`review-text-${review.title}`} className={styles.modalText}>
                    {review.text}
                </article>
            </div>
        </aside>
    );
};

export default PublicJournalEntryModal;
