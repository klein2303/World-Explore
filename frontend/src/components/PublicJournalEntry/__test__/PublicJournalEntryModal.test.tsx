import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import PublicJournalEntryModal from "../PublicJournalEntryModal";
import { readReviewType } from "../../../types/JournalType";

describe("PublicJournalEntryModal Component", () => {
    const mockOnClose = vi.fn();
    const mockReview: readReviewType = {
        title: "Wonderful Journey",
        date: "2024-11-20",
        rating: 4,
        text: "This experience was truly amazing!",
        public: true,
        journal: {
            profile: {
                username: "traveler123",
            },
        },
    };

    beforeEach(() => {
        // Clear the mock function's call history before each test
        mockOnClose.mockClear();
    });

    it("renders correctly and matches snapshot", () => {
        const { asFragment } = render(<PublicJournalEntryModal review={mockReview} onClose={mockOnClose} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("displays the correct review details", () => {
        render(<PublicJournalEntryModal review={mockReview} onClose={mockOnClose} />);

        // Check that the title, date, rating, and username are rendered
        expect(screen.getByText(mockReview.title)).toBeInTheDocument();
        expect(screen.getByText(mockReview.date)).toBeInTheDocument();
        expect(screen.getByText(`${mockReview.rating}/5`)).toBeInTheDocument();
        expect(screen.getByText(mockReview.journal.profile.username)).toBeInTheDocument();
        expect(screen.getByText(mockReview.text)).toBeInTheDocument();
    });

    it("closes the modal when the close button is clicked", () => {
        render(<PublicJournalEntryModal review={mockReview} onClose={mockOnClose} />);

        const closeButton = screen.getByLabelText("Close modal");
        fireEvent.click(closeButton);

        // Ensure the onClose callback is called
        expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it("closes the modal when the Escape key is pressed", () => {
        render(<PublicJournalEntryModal review={mockReview} onClose={mockOnClose} />);

        fireEvent.keyDown(document, { key: "Escape", code: "Escape" });

        // Ensure the onClose callback is called
        expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it("closes the modal when clicking on the backdrop", () => {
        render(<PublicJournalEntryModal review={mockReview} onClose={mockOnClose} />);

        const backdrop = screen.getByRole("dialog");
        fireEvent.click(backdrop);

        // Ensure the onClose callback is called
        expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    it("does not close the modal when clicking inside the modal content", () => {
        render(<PublicJournalEntryModal review={mockReview} onClose={mockOnClose} />);

        const modalContent = screen.getByText(mockReview.title).closest("div");
        fireEvent.click(modalContent!);

        // Ensure the onClose callback is not called
        expect(mockOnClose).not.toHaveBeenCalled();
    });
});
