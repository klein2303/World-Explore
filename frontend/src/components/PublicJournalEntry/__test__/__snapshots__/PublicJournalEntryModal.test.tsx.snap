// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PublicJournalEntryModal Component > renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <aside
    aria-describedby="review-text-This experience was truly amazing!"
    aria-labelledby="review-title-Wonderful Journey"
    aria-modal="true"
    class="_modalBackdrop_62482e"
    role="dialog"
  >
    <div
      class="_modalContent_62482e"
      tabindex="-1"
    >
      <header>
        <button
          aria-label="Close modal"
          class="_closeButton_62482e"
        >
          Ã—
        </button>
        <h3
          class="_modalTitle_62482e"
          id="review-title-Wonderful Journey"
        >
          Wonderful Journey
        </h3>
        <p
          aria-label="Date range: 2024-11-20"
          class="_modalDate_62482e"
        >
          2024-11-20
        </p>
      </header>
      <section
        aria-labelledby="modal-rating modal-reviewer"
        class="_modalInfo_62482e"
      >
        <p
          aria-description="Rating: 4 out of 5 stars"
          class="_modalRating_62482e"
          id="modal-rating"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-star _starIcon_62482e"
            data-icon="star"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
              fill="currentColor"
            />
          </svg>
          4/5
        </p>
        <p
          class="_user_62482e"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H322.8c-3.1-8.8-3.7-18.4-1.4-27.8l15-60.1c2.8-11.3 8.6-21.5 16.8-29.7l40.3-40.3c-32.1-31-75.7-50.1-123.9-50.1H178.3zm435.5-68.3c-15.6-15.6-40.9-15.6-56.6 0l-29.4 29.4 71 71 29.4-29.4c15.6-15.6 15.6-40.9 0-56.6l-14.4-14.4zM375.9 417c-4.1 4.1-7 9.2-8.4 14.9l-15 60.1c-1.4 5.5 .2 11.2 4.2 15.2s9.7 5.6 15.2 4.2l60.1-15c5.6-1.4 10.8-4.3 14.9-8.4L576.1 358.7l-71-71L375.9 417z"
            />
          </svg>
          traveler123
        </p>
      </section>
      <article
        class="_modalText_62482e"
        id="review-text-Wonderful Journey"
      >
        This experience was truly amazing!
      </article>
    </div>
  </aside>
</DocumentFragment>
`;
